### PR TITLE
feat: add recipe semantic validation

### PR DESCRIPTION
**Pull request by nullPointer373**
**Issue #, if available:**

**Description of changes:**
Implemented semantic validation for recipe files and included corresponding unit tests. Introduced the capability to parse validation errors if any issues arise in the recipe's structure or input options. Additionally, the validation flow has been integrated into the build stage, ensuring that user input recipes are thoroughly validated before proceeding with the transformation into the build directory.

**Why is this change necessary:**
This enhancement is crucial to guarantee the integrity of recipe data and to prevent downstream issues during processing and deployment.

**How was this change tested:**
Added unit tests to validate the functionality, and all unit tests (including the new ones) and integration tests passed successfully.

**Any additional information or context required to review the change:**

**Checklist:**
- [ ] Updated the README if applicable
- [x] Updated or added new unit tests
- [ ] Updated or added new integration tests
- [ ] Updated or added new end-to-end tests
- [ ] If your code makes a remote network call, it was tested with a proxy
- [ ] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-kevinrickard** [comment]
Could you post example output for of a couple use cases, e.g. 
1. A valid recipe used in gdk component build, 
2. An invalid recipe used in gdk component build ? 

This would help me see some of the new outputs in context

File: gdk/common/RecipeValidator.py [hunk 1]
```diff
@@ -43,20 +49,37 @@ def __init__(self, recipe_source):
...
         self.recipe_data = self._load_recipe()
-        # TODO: add the semantic validation logic
-        pass
+        recipe_schema = utils.get_static_file_path(consts.user_input_recipe_schema_file)
+        with open(recipe_schema, 'r') as schema_file:
+            schema = json.load(schema_file)
+        logging.debug("Validating the recipe file.")
+        data = self._convert_keys_to_lowercase(self.recipe_data)
+        try:
+            jsonschema.validate(data, schema)
```
Comment (aws-kevinrickard) [inline gdk/common/RecipeValidator.py:71]: I presume this works for both yaml and json recipes? If so, nice!

**Review by aws-kevinrickard — COMMENTED** [review]
Overall looks good, but have these small questions

**Comment by nullPointer373** [comment]
> 

For a valid recipe, the output is:
[2023-08-16 09:52:05] INFO - Building the component 'com.example.PythonHelloWorld' with the given project configuration.
[2023-08-16 09:52:05] INFO - Using 'zip' build system to build the component.
[2023-08-16 09:52:05] WARNING - This component is identified as using 'zip' build system. If this is incorrect, please exit and specify custom build command in the 'gdk-config.json'.
[2023-08-16 09:52:05] INFO - Copying over the build artifacts to the greengrass component artifacts build folder.
[2023-08-16 09:52:05] INFO - Updating artifact URIs in the recipe.

For an invalid recipe, the output will be:
- If the recipe contains syntax error (wrong indentation in yaml file):

[2023-08-16 09:54:55] INFO - Building the component 'com.example.PythonHelloWorld' with the given project configuration.
[2023-08-16 09:54:55] INFO - Using 'zip' build system to build the component.
[2023-08-16 09:54:55] WARNING - This component is identified as using 'zip' build system. If this is incorrect, please exit and specify custom build command in the 'gdk-config.json'.
[2023-08-16 09:54:55] ERROR - Syntax error when parsing the recipe file: /Users/yifanzyf/Desktop/HelloWorld/recipe.yaml
[2023-08-16 09:54:55] ERROR - 

while parsing a block mapping
  in "<unicode string>", line 2, column 1:
    recipeFormatVersion: "2020-01-25"
    ^
expected <block end>, but found '<block mapping start>'
  in "<unicode string>", line 6, column 2:
     ComponentPublisher: "{COMPONENT_ ... 
     ^
[2023-08-16 09:54:55] INFO - This might be caused by one of the following reasons: 
[2023-08-16 09:54:55] INFO -     unexpected characters or tokens at the end of the line
[2023-08-16 09:54:55] INFO -     incorrect indentation
[2023-08-16 09:54:55] INFO -     missing closing quotes on the line above
[2023-08-16 09:54:55] INFO -     unmatched opening and closing quotes
[2023-08-16 09:54:55] INFO -     missing mandatory space after the character '-' in the line above
[2023-08-16 09:54:55] INFO - Please review the overall YAML syntax and resolve any issues.
[2023-08-16 09:54:55] ERROR - Failed to build the component with the given project configuration.

- If the recipe contains semantic error (wrong value type for a property):

[2023-08-16 09:57:46] INFO - Building the component 'com.example.PythonHelloWorld' with the given project configuration.
[2023-08-16 09:57:46] INFO - Using 'zip' build system to build the component.
[2023-08-16 09:57:46] WARNING - This component is identified as using 'zip' build system. If this is incorrect, please exit and specify custom build command in the 'gdk-config.json'.
[2023-08-16 09:57:46] ERROR - Error when validating recipe file: 27 is not of type 'string'
[2023-08-16 09:57:46] ERROR - Validation failed for 'type: string'
[2023-08-16 09:57:46] INFO - On instance: 'componentversion: 27'
[2023-08-16 09:57:46] INFO - This validation error may be due to: 
[2023-08-16 09:57:46] INFO -     Instance type does not match the specified type.
[2023-08-16 09:57:46] INFO - To address this issue, consider the following steps: 
[2023-08-16 09:57:46] INFO -     Please check the instance type and ensure it matches the expected type.
[2023-08-16 09:57:46] ERROR - Failed to build the component with the given project configuration.

File: gdk/common/RecipeValidator.py (hunk 1 above)
Comment (nullPointer373) [reply gdk/common/RecipeValidator.py:71]: This works for both yaml and json recipes once the recipe data is read

**Review by saranyailla — APPROVED** [review]

**Review by aws-kevinrickard — APPROVED** [review]

**Pull request merged**